### ISSUE TITLE
Validate CSP for Blazor WASM

### ISSUE THREAD

**Issue by TanayParikh** [dotnet/aspnetcore#37787]
Runtime made the necessary changes that `unsafe-eval` may no longer be required for Blazor WASM: https://github.com/dotnet/runtime/issues/59416

Note, there's also a Chrome bug which may come into play here (WebAssembly/content-security-policy#7)

This issue tracks validating this fix, and the associated documentation updates.

Original CSP Issue: https://github.com/dotnet/aspnetcore/issues/34428

**Comment by TanayParikh** [comment]
Unfortunately this is still an issue due to the [Chrome bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1169592):

<img width="1525" alt="Screen Shot 2021-12-08 at 12 30 18 PM" src="https://user-images.githubusercontent.com/14852843/145279811-e12d2a07-19b1-490c-a3ef-9d689091cd24.png">

This resource helps detect if the chrome issue is resolved (WASM related, not blazor specific): https://s3.amazonaws.com/webassembly-chrome-csp/csp_test.html

Others are having the same issue as well: https://github.com/vector-im/element-web/issues/12262

**Comment by mingyaulee** [comment]
@TanayParikh based on my testing with dotnet.7.0.0-preview.3.22175.4 the runtime still requires `unsafe-eval`, mainly due to the evaluation of string in the following line:

https://github.com/dotnet/runtime/blob/14584c60b41ddc361442539f78bc3d54d3ab3ea2/src/mono/wasm/runtime/method-binding.ts#L113

The original issue was closed and locked so I am hoping you might be able to reopen it.

**Comment by TanayParikh** [comment]
> original issue

Can you please provide a link.

**Comment by mingyaulee** [comment]
I was referring to the issue that you have created and linked to from here.
[dotnet/runtime#59416 Make `mono` CSP Compliant](https://github.com/dotnet/runtime/issues/59416)

**Comment by TanayParikh** [comment]
Ah so it's over on the runtime repo. Thanks for clarifying. Please open up a new issue with the line you pointed out, on dotnet/runtime, and crosslink https://github.com/dotnet/runtime/issues/59416 and  https://github.com/dotnet/aspnetcore/issues/37787.

**Comment by SteveSandersonMS** [comment]
We should re-check if this is still the case. Since this issue was posted, Chromium has added support for a new CSP flag, `wasm-unsafe-eval`, which is intended for this case. It might now be possible to run without `unsafe-eval`, by adding `wasm-unsafe-eval` instead: https://chromestatus.com/feature/5499765773041664

**Comment by TanayParikh** [comment]
Unfortunately `wasm-unsafe-eval` doesn't look to be sufficient in this case:

<img width="1040" alt="image" src="https://user-images.githubusercontent.com/14852843/172720127-7371f3cd-845a-4e9e-95fe-b9ea983f4fca.png">

It's calling out https://github.com/dotnet/runtime/blob/637f8f241146e467d6a603394d63ff716838c0c6/src/mono/wasm/runtime/method-binding.ts#L81-L115 specifically. 

cc/ @radical any ideas what may be going wrong here?

**Comment by radical** [comment]
https://github.com/WebAssembly/content-security-policy/blob/master/proposals/CSP.md#proposed-policy-behavior

```
If the wasm-unsafe-eval source keyword is used, and the unsafe-eval keyword is not present, then WebAssembly modules may be compiled and instantiated.

   One advantage of this is that a website can permit WebAssembly modules to be used without also enabling JavaScript's eval keyword.
```

IIUC, this means that `wasm-unsafe-eval` will allow `WebAssembly.compile`, and friends to work, but *not* the general `eval` in JS. Do correct me if I'm wrong. So, then the situation is unchanged.

cc @pavelsavara @lewing

**Comment by pavelsavara** [comment]
Yes, we need to get rid of eval in the JS code first. Also see https://github.com/dotnet/runtime/issues/68374

**Comment by veikkoeeva** [comment]
I was looking into this case also. Maybe relevant is recent discussion regarding WASM loading in extensions and v3 manifest at https://bugs.chromium.org/p/chromium/issues/detail?id=1318922 (made this comment also in the previously linked thread, found it first).

**Comment by KieranDevvs** [comment]
> 

I'm also wanting to use Blazor in a chrome extension but this is blocking me.

**Comment by davidchavezgrant** [comment]
> I'm also wanting to use Blazor in a chrome extension but this is blocking me.

Same here.

**Comment by veikkoeeva** [comment]
Should https://github.com/dotnet/runtime/pull/74441 close this also?

**Comment by pavelsavara** [comment]
Blazor in Net7 is using legacy JavaScript interop, which is not CSP safe. We need to migrate Blazor to use the new JS interop with `[JSImportAttribute]` to make it CSP friendly.

**Comment by KieranDevvs** [comment]
> Just to show that its not a rare or low-impact issue. 

I second that.

**Comment by ctigrisht** [comment]
we need this fixed asap, cannot realistically use blazor wasm without this

**Comment by andrewrondeau-optirtc** [comment]
I need to point out that this is a very important thing to fix in order for Blazor to be considered mature.

Specifically, "eval()" in Javascript is considered insecure, but Blazor essentially runs via "eval()."

It's quite critical to fix this, in the long-term, for a Blazor-powered site to be considered safe from XSS attacks.
